activerecord/test/models/bird.rb

```diff
@@ -16,4 +16,9 @@ class Bird < ActiveRecord::Base
   def cancel_save_callback_method
     throw(:abort)
   end
+
+  attr_accessor :total_count
+  after_initialize do
+    self.total_count = Bird.count
```

Comment:
> Would not this cause breaking changes?

Before this PR #35186, the `self.total_count = Bird.count` in `after_initialize` and the `Bird.find_by!(name: "canary")` are affected by `Bird.where(color: "green").scoping { }` in `Bird.where(color: "green").create`, so `parrot.total_count` will be 0 and `Bird.find_by!(name: "canary")` will raise `RecordNotFound`.

```ruby
canary = Bird.create!(color: "yellow", name: "canary")
parrot = Bird.where(color: "green").create do |bird|
  bird.name = "parrot"
  assert_equal canary, Bird.find_by!(name: "canary")
end
assert_equal 1, parrot.total_count
```

If anyone depends on the behavior, it is affected by this change.
But I think the leaking scope is just a bug, probably most people use a workaround `Bird.unscoped.count` and `Bird.unscoped.find_by!(name: "canary")` for a long time.

Reply:
:+1

Reply:
@kamipo Thank you for the fix :-)

Reply:
I'm still concerned about this breaking change. If people is relying on this behavior they may start to leak information silently like happened some years ago in the GitHub incident.

Is there any way we can detect people is relying on this behavior and show them a warning?

@tenderlove @matthewd @jeremy what do you think?

Reply:
Please correct me if I'm wrong, but if the scope is leaking, then the scope in `after_initialize` etc would depend on the context in which the object is created.  People depending on this behavior would have to have exactly one code path for initialization, otherwise they would have to do `unscoped` because the scope would be different each time.

Could we set the scope object to something that raises an exception if `unscoped` *isn't* called?

Reply:
Since `scoping` is public API, people may have a usage like the case (1).
In the case (1), initialization block and callbacks are affected by the `Bird.where(color: "yellow").scoping`.
I made the case (2) working the same with the case (1) in this PR.

We could set the exception scope object for the case (2), it will make the case (2) working different back with the case (1) though...

```ruby
canary = Bird.create!(color: "yellow", name: "canary")

# (1)
Bird.where(color: "yellow").scoping do
  parrot = Bird.create!(color: "green") do |bird|
    bird.name = "parrot"
    assert_equal canary, Bird.find_by!(name: "canary")
  end
  assert_equal 1, parrot.total_count
end

# (2)
Bird.where(color: "yellow").scoping do
  parrot = Bird.rewhere(color: "green").create! do |bird|
    bird.name = "parrot"
    assert_equal canary, Bird.find_by!(name: "canary")
  end
  assert_equal 1, parrot.total_count
end
```